src/controller/OperationalCredentialsDelegate.h

```diff
@@ -0,0 +1,99 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/util/basic-types.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <support/DLLUtil.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+namespace Controller {
+
+/// Callbacks for CHIP operational credentials generation
+class DLL_EXPORT OperationalCredentialsDelegate
+{
+public:
+    virtual ~OperationalCredentialsDelegate() {}
+
+    /**
+     * @brief
+     *   This function generates an operational certificate for the given node.
+     *   The API generates the certificate in X.509 DER format.
+     *
+     *   The delegate is expected to use the certificate authority whose certificate
+     *   is returned in `GetIntermediateCACertificate()` or `GetRootCACertificate()`
+     *   API calls.
+     *
+     * @param[in] nodeId       Node ID of the target device.
+     * @param[in] fabricId     Fabric ID for which the credentials are being generated.
+     * @param[in] csr          Certificate Signing Request from the node in DER format.
+     * @param[in] csrLength    The length of Certificate Signing Request.
+     * @param[in] serialNumber Serial number to assign to the new certificate.
+     * @param[in] certBuf      The API will fill in the generated cert in this buffer. The buffer is allocated by the caller.
+     * @param[in] certBufSize  The size of certBuf buffer.
+     * @param[out] outCertLen  The size of the actual certificate that was written in the certBuf.
+     *
+     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     */
+    virtual CHIP_ERROR GenerateNodeOperationalCertificate(NodeId nodeId, FabricId fabricId, const uint8_t * csr, size_t csrLength,
+                                                          int64_t serialNumber, uint8_t * certBuf, uint32_t certBufSize,
+                                                          uint32_t & outCertLen) = 0;
+
+    /**
+     * @brief
+     *   This function returns the intermediate certificate authority (ICA) certificate corresponding to the
+     *   provided fabric ID. Intermediate certificate authority is optional. If the controller
+     *   application does not require ICA, this API call will return `CHIP_ERROR_NOT_IMPLEMENTED`.
+     *
+     *   The returned certificate is in X.509 DER format.
+     *
+     * @param[in] fabricId    Fabric ID for which the certificate is being requested.
+     * @param[in] certBuf     The API will fill in the cert in this buffer. The buffer is allocated by the caller.
+     * @param[in] certBufSize The size of certBuf buffer.
+     * @param[out] outCertLen The size of the actual certificate that was written in the certBuf.
+     *
+     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     *         CHIP_ERROR_NOT_IMPLEMENTED is not a critical error. It indicates that ICA is not needed.
```

Comment:
Can we define a new error that say 'CHIP_ERROR_INTERMEDIATE_CA_NOT_REQUIRED' that would make it clear that it is not an issue?

Alternatively we could have outCertLen have some magic value (0?) and success signify no certificate. NOT_IMPLEMENTED sounds an odd return for 'OK'